File: Python/Snippet.py
{
	// Place your snippets for python here. Each snippet is defined under a snippet name and has a prefix, body and
	// description. The prefix is what is used to trigger the snippet and the body will be expanded and inserted. Possible variables are:
	// $1, $2 for tab stops, $0 for the final cursor position, and ${1:label}, ${2:another} for placeholders. Placeholders with the
	// same ids are connected.
	// Example:
	"Print to console": {
		"prefix": "te",
		"body": [
			// 		"console.log('$1');",
			// 		"$2"
			"from os import read, write, fstat; from sys import stdin, stdout; from io import BytesIO, IOBase; BS = 8192",
			"class FastIO(IOBase):",
			"\tnl = 0",
			"\tdef __init__(self, file): self._file = file; self._fd = file.fileno(); self.buffer = BytesIO(); self.writable = \"x\" in file.mode or \"r\" not in file.mode; self.write = self.buffer.write if self.writable else None",
			"\tdef read(self):",
			"\t\twhile True:",
            "\t\t\tb = read(self._fd, max(fstat(self._fd).st_size, BS))",
            "\t\t\tif not b: break",
            "\t\t\tptr = self.buffer.tell(); self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)",
			"\t\tself.nl = 0; return self.buffer.read()",
			"\tdef readline(self):",
			"\t\twhile not self.nl: b = read(self._fd, max(fstat(self._fd).st_size, BS)); self.nl = b.count(b\"\\n\") + (not b); ptr = self.buffer.tell(); self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)",
			"\t\tself.nl -= 1; return self.buffer.readline()",
			"\tdef flush(self):",
			"\t\tif self.writable: write(self._fd, self.buffer.getvalue()); self.buffer.truncate(0), self.buffer.seek(0)",
			"class IOWrapper(IOBase):",
			"\tdef __init__(self, file): self.buffer = FastIO(file); self.flush = self.buffer.flush; self.writable = self.buffer.writable; self.write = lambda s: self.buffer.write(s.encode(\"ascii\")); self.read = lambda: self.buffer.read().decode(\"ascii\"); self.readline = lambda: self.buffer.readline().decode(\"ascii\")",
			"stdin, stdout = IOWrapper(stdin), IOWrapper(stdout); input = lambda: stdin.readline().rstrip(\"\\r\\n\")",
			"def ck(*A): from sys import stderr; from inspect import currentframe; stderr.write(\", \".join([f\"{t} = {v}\" for t, v in zip([t for t, v in currentframe().f_back.f_locals.items() if v in A], A)]) + '\\n')",
			"inp = lambda: (c := stdin.readline().split(), int(c[0]) if len(c) == 1 else [int(x) for x in c])[1]",
			"def pn(c, n): [stdout.write(str(c[i]) + ' ') for i in range(n)]; stdout.write('\\n')",
			"def ols(ls): [stdout.write(str(x) + ' ') for x in ls]; stdout.write('\\n')",
			"def ps(b): stdout.write(\"YES\\n\" if b else \"NO\\n\")",
			"def pc(t): stdout.write(\"Case \" + str(t) + \": \")",
			"def out(n): stdout.write(str(n) + '\\n')",
			"M, N = 1000000007, 200005",
			"",
			"def Test(tc):",
			"\tn = inp()",
			"\tls = inp()",
			"\t$0",
			"",
			"if __name__ == \"__main__\":",
			"\tt = tc = 1",
			"\ttc = inp()",
			"\twhile t <= tc: Test(t); t += 1",
		],
	// 	"description": "Log output to console"
	// }
}

File: CPP/Snippet.cpp
{
	// Place your snippets for cpp here. Each snippet is defined under a snippet name and has a prefix, body and
	// description. The prefix is what is used to trigger the snippet and the body will be expanded and inserted. Possible variables are:
	// $1, $2 for tab stops, $0 for the final cursor position, and ${1:label}, ${2:another} for placeholders. Placeholders with the
	// same ids are connected.
	// Example:
	"Print to console": {
		"prefix": "te",
		"body": [
			// 		"console.log('$1');",
			// 		"$2"
			"#include <bits/stdc++.h>",
			"using namespace std;",
			"#define ll long long",
			"#define pe(c) for (auto &e : c) cout << e << ' '; cout << '\\n'",
			"#define ps(b) cout << (b ? \"YES\" : \"NO\") << '\\n'",
			"#ifdef LOCAL",
			"#include \"def.h\"",
			"#else""#define ck(...)",
			"#endif",
			"const ll M = 1e9 + 7, N = 2e5 + 5;",
			"",
			"void test(int tc) {",
			"\tll n = 0, m = 0, a = 0, b = 0, c = 0, d = 0, i = 0, j = 0, k = 0, q = 0;",
			"\tcin >> n;",
			"\tvector<ll> ar(n); for (i = 0; i < n; ++i) { cin >> ar[i]; }",
			"\t$0",
			"\tcout << '\\n';",
			"}",
			"",
			"signed main() {",
			"\tcin.tie(0)->sync_with_stdio(0); cin.exceptions(cin.failbit | cin.badbit);",
			"\tint tc = 0, t = 1;",
			"\tcin >> t;",
			"\twhile (tc < t) test(++tc);",
			"\treturn 0;",
			"}",
		],
		"description": "Log output to console",
	},

	"Print to console1": {
		"prefix": "mi",
		"body": [
			"template <const int64_t MOD>",
			"struct modint {",
			"\tint64_t val; modint() = default; modint(int64_t value) : val(value % MOD) { if (val < 0) val += MOD; }",
			"\tinline modint<MOD> operator+(modint<MOD> other) const { int64_t c = this->val + other.val; return modint<MOD>(c >= MOD ? c - MOD : c); }",
			"\tinline modint<MOD> operator-(modint<MOD> other) const { int64_t c = this->val - other.val; return modint<MOD>(c < 0 ? c + MOD : c); }",
			"\tinline modint<MOD> operator*(modint<MOD> other) const { int64_t c = (__int128_t)this->val * other.val % MOD; return modint<MOD>(c < 0 ? c + MOD : c); }",
			"\tinline modint<MOD> &operator+=(modint<MOD> other) { this->val += other.val; if (this->val >= MOD) this->val -= MOD; return *this; }",
			"\tinline modint<MOD> &operator-=(modint<MOD> other) { this->val -= other.val; if (this->val < 0) this->val += MOD; return *this; }",
			"\tinline modint<MOD> &operator*=(modint<MOD> other) { this->val = (__int128_t)this->val * other.val % MOD; if (this->val < 0) this->val += MOD; return *this; }",
			"\tinline modint<MOD> operator-() const { return modint<MOD>(this->val ? MOD - this->val : 0); }",
			"\tmodint<MOD> pow(uint64_t k) const { modint<MOD> x = *this, y = 1; for (; k; k >>= 1) { if (k & 1) y *= x; x *= x; } return y; }",
			"\tmodint<MOD> inv() const { return pow(MOD - 2); }",
			"\tinline modint<MOD> operator/(modint<MOD> other) const { return *this * other.inv(); }",
			"\tinline modint<MOD> operator/=(modint<MOD> other) { return *this *= other.inv(); }",
			"\tinline bool operator==(modint<MOD> other) const { return val == other.val; }",
			"\tinline bool operator!=(modint<MOD> other) const { return val != other.val; }",
			"\tinline bool operator<(modint<MOD> other) const { return val < other.val; }",
			"\tinline bool operator>(modint<MOD> other) const { return val > other.val; }",
			"};",
			"template <int64_t MOD> modint<MOD> operator*(int64_t val, modint<MOD> n) { return modint<MOD>(val) * n; }",
			"template <int64_t MOD> modint<MOD> operator*(int32_t val, modint<MOD> n) { return modint<MOD>(val % MOD) * n; }",
			"template <int64_t MOD> istream &operator>>(istream &in, modint<MOD> &n) { int64_t val; in >> val; n = modint<MOD>(val); return in; }",
			"template <int64_t MOD> ostream &operator<<(ostream &out, modint<MOD> n) { return out << n.val; }",
			"using mint = modint<M>;",
		],
		// "description": "Log output to console",
	}
}